_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,402 @@
+  `
+})
+export class VideoComponent {
+
+  // pretend this is from an external source
+  public videoId = '12345';
+  public safeLinky!: SafeResourceUrl;
+
+  constructor(private sanitizer: DomSanitizer) {
+    this.safeLinky = sanitizer.bypassSecurityTrustResourceUrl(`//videolink/embed/${this.videoId}`)
+  }
+}
+```
+
+Now you'll be able to show trailers of the K-Dramas on your site in an `iframe` in a much safer way.
+
+> It can't be stressed enough. Ensure you trust the values before bypassing the out-of-the-box security protections you get from Angular!
+
+Great! So we're done! Not quite. There are a couple of things to note.
+
+### Use Ahead-of-time (AOT) compilation for extra security
+
+Angular's AOT compilation has extra security measures for injection attacks like XSS. AOT compilation is highly recommended for production code and has been the default compilation method since Angular v9. Not only is it more secure, but it also increases performance.
+
+On the flip side, the other form of compilation is Just-in-time (JIT). JIT was the default for older versions of Angular. JIT compiles code for the browser on the fly, and this process skips Angular's built-in security protection, so stick with using AOT.
+
+### Don't concatenate strings to construct templates
+
+Angular trusts template code and only escapes values defined in the template using interpolation. So if you attempt something clever to circumvent the more common forms of defining the template for a component, you won't be protected. 
+
+For example, you won't have Angular's built-in protections if you try to dynamically construct templates combining HTML with data using string concatenation or have an API whip up a payload with a template you somehow inject into the app. Your clever hacks with dynamic components might cause you security woes. 
+
+### Beware of constructing DOM elements without using Angular templates
+
+Any funny business you might try using `ElementRef` or `Renderer2` is the perfect way to cause security woes for yourself. For example, you can pwn yourself if you try to do something like this.
+
+```ts
+@Component({
+  selector: 'app-yikes',
+  template: `
+    <div #whydothis></div>
+  `
+})
+export class YikesComponent implements AfterViewInit {
+
+  @ViewChild('whydothis') public el!: ElementRef<HTMLElement>;
+  
+  // pretend this is from an external source
+  public attack = '<img src=1 onerror="alert(\'YIKES!\')"';
+
+  constructor(private renderer: Renderer2) { }
+
+  public ngAfterViewInit(): void {
+    
+    // danger below!
+    this.el.nativeElement.innerHTML = this.attack;
+    this.renderer.setProperty(this.el.nativeElement, 'innerHTML', this.attack);
+  }
+}
+```
+
+Something like this might be tempting in a fancy custom directive, but think again! Besides, interacting directly with the DOM like this isn't the best practice in Angular, despite any security issues it might have. You should always prefer creating and using Angular templates.
```

Comment:
"despite" - i'm wondering if you mean "even beyond the security issues it might in the sentence:

 "... isn't the best practice in Angular, despite any security issues it might have."   
 
For "You should always prefer creating and using Angular templates." - Could you say: "Always use and create Angular templates when you can." - it seems just as opinionated, and more direct.

Reply:
Yes to both. I'll update, thanks!